### PR TITLE
Do not process main thread queued calls while returning a worker to the pool

### LINKED ISSUE #12970: Flaky "Internal error!" and other exceptions when using threads
Reproducible using this simple program, although with a big number of iterations:

```
#include <thread>
int main() {
  for (int i = 0; i < 1000000; ++i)
    std::thread([](){}).join();
}
```

Compilation flags (using Emscripten 2.0.6):
```
emcc main.cc -s INITIAL_MEMORY=67108864 -pthread -s PROXY_TO_PTHREAD=1 -s EXIT_RUNTIME=1
```

This program typically crashes with an exception after some number of iterations (sometimes 100, sometimes as big as 30000), e.g.:

```
Internal error!
Thrown at:
    at a.out.js:151:7
    at emit (events.js:315:20)
    at internal/process/execution.js:165:25
```

or (when running in Chrome, the stack trace is fuller for whatever reason);

```
spawnThread	@	a.out.js:3125
_pthread_create	@	a.out.js:3349
(anonymous)	@	a.out.wasm:0x87b0
emscripten_current_thread_process_queued_calls	@	a.out.wasm:0x83e0
emscripten_main_thread_process_queued_calls	@	a.out.wasm:0x941b
(anonymous)	@	a.out.wasm:0x7f2b
(anonymous)	@	a.out.wasm:0x7ff8
(anonymous)	@	a.out.wasm:0x7527
(anonymous)	@	a.out.wasm:0x73f3
free	@	a.out.wasm:0x63fd
(anonymous)	@	a.out.js:1615
freeThreadData	@	a.out.js:2039
returnWorkerToPool	@	a.out.js:2047
worker.onmessage	@	a.out.js:2101
```

or

```
Pthread aborting at Error
    at abort (eval at globalEval (a.out.worker.js:200:10), <anonymous>:1630:70)
    at _abort (eval at globalEval (a.out.worker.js:200:10), <anonymous>:2410:7)
    at dlfree (wasm-function[264]:0x76b0)
    at operator delete(void*) (wasm-function[99]:0x208b)
    at std::__2::default_delete<std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0> >::operator()(std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0>*) const (wasm-function[91]:0x1ed5)
    at std::__2::unique_ptr<std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0>, std::__2::default_delete<std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0> > >::reset(std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0>*) (wasm-function[59]:0x1633)
    at std::__2::unique_ptr<std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0>, std::__2::default_delete<std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0> > >::~unique_ptr() (wasm-function[46]:0x117c)
    at void* std::__2::__thread_proxy<std::__2::tuple<std::__2::unique_ptr<std::__2::__thread_struct, std::__2::default_delete<std::__2::__thread_struct> >, main::$_0> >(void*) (wasm-function[36]:0xe42)
    at Object.dynCall (eval at globalEval (a.out.worker.js:200:10), <anonymous>:1945:33)
    at onmessage (a.out.worker.js:124:39)
```

The exceptions are seen regardless of how it's executed: Node (12.18.1) or Chrome (87.0.4280.66).

I believe I also see the same errors in a non-synthetic real-world program (with much smaller number of threads but a big amount of other functionality).

### PR DESCRIPTION
**Pull request by kripken**
This adds a utility function for not running queued calls in a scope. The method
of avoiding them is slightly hackish, by temporarily re-registering the main
thread as if it were not the main runtime thread, but as it's limited to that
utility function, I think it's reasonable, and simpler than adding a new
mechanism. But open to ideas.

Fixes #12970

### CODE REVIEW

**Review by sbc100 — COMMENTED** [review]
This seems a little hacky to me.

Firstly I don't understand why doing this does't cause the futex code to start using `Atomics.wait` rather then busy waiting.   Looking at code it looks like maybe we never use the `Atomics.wait` on the web, even on workers which seems strange.  Anyway, setting this bit could/should trigger the use of `Atomics.wait` so it seems like the wrong thing to do.

How about doing something more precise such as temporarily setting `emscripten_futex_wait.do_not_run_callback_while_waiting = True`?



Could we not have a separate flag specifically for the mutux here?

**Comment by sbc100** [comment]
> Firstly I don't understand why doing this does't cause the futex code to start using `Atomics.wait` rather then busy waiting. Looking at code it looks like maybe we never use the `Atomics.wait` on the web, even on workers which seems strange. Anyway, setting this bit could/should trigger the use of `Atomics.wait` so it seems like the wrong thing to do.
>

Ah... it looks like the use of `Atomics.wait` is guarded by `ENVIRONMENT_IS_WEB` which I hadn't realized was not set on workers... this seems odd.. surly it would make sense to check "is_main_browser_thread"?    And wouldn't it make more sense to set  ENVIRONMENT_IS_WEB on all the workers too?

**Comment by kripken** [comment]
> Could we not have a separate flag specifically for the mutex here?

We could, but it would need to live on the C++ side, so it would sort of duplicate the `init` function I'm reusing here. But if you think it's worth the extra code to reduce hackishness, sgtm?

> Ah... it looks like the use of Atomics.wait is guarded by ENVIRONMENT_IS_WEB which I hadn't realized was not set on workers... this seems odd.. surely it would make sense to check "is_main_browser_thread"? 

That should be the same check. `ENVIRONMENT_IS_WEB` is set exactly on the main browser thread. (But it is just available in JS.)

> And wouldn't it make more sense to set ENVIRONMENT_IS_WEB on all the workers too?

WEB and WORKER are different environments in our modelling. Only one of the `ENVIRONMENT_IS_` are defined as true, normally (but I think node workers are an exception as they get both NODE and WORKER - we may want to revisit that).

**Comment by sbc100** [comment]
Looking at `emscripten_futex_wait` implementation it looks like the logic is basically:

```
if (not ENV_IS_WEB)
   call Atomics.wait();
else 
   busy loop && call `_emscripten_main_thread_process_queued_calls`
```

So even with you change I think `_emscripten_main_thread_process_queued_calls` still get called in a loop (although that function does have an early out).   How about a flag on the JS side the causes the busy loop in emscripten_futex_wait to skip over `_emscripten_main_thread_process_queued_calls`?  It seems cleaner to me than calling emscripten_thread_init().

**Comment by kripken** [comment]
> So even with you change I think _emscripten_main_thread_process_queued_calls still get called in a loop

That is a separate issue from this PR. On the main thread on the web, we must call it in a loop since we can't block. That is regardless of whether we run queued calls or not.

>  How about a flag on the JS side the causes the busy loop in emscripten_futex_wait to skip over _emscripten_main_thread_process_queued_calls?

That is not enough. That function is also called from C code (`__wait, __timedwait`).

But I agree the original code here was too hackish. PTAL now, I added a global in C that guards the function.

File: emcc.py [hunk 1]
```diff
@@ -1523,6 +1523,7 @@ def default_setting(name, new_default):
         '__emscripten_main_thread_futex',
         '__emscripten_thread_init',
         '_emscripten_current_thread_process_queued_calls',
+        '__emscripten_allow_main_runtime_queued_calls',
```
Comment (sbc100) [inline emcc.py:1526]: It occurs to me that is kind of sad that these functions that are exported only for internal use probably can't be minimized?   Is that right?   (Or do use the user_exported_functions trick to know that they actually don't need to be fully exported?)   Unrelated to this PR of course.

**Review by sbc100 — APPROVED** [review]
Much nicer.

File: emcc.py (hunk 1 above)
Comment (kripken) [reply emcc.py:1526]: This should be ok. First, the names are minified into `a`, `b` etc. Second, metadce can remove the export if it's not used at all, as it's not a user-requested export.

**Pull request merged**